File: csharp/word-search/WordSearch.cs
using Xunit.Internal;

public class WordSearch
{
    string[] grid;
    /// <summary>
    /// Constructor. Saves the grid for later word search calls
    /// </summary>
    /// <param name="grid">A string where each row is separated
    /// by a '\n' character and the rows are the values inbetween.
    /// It is assumed that all rows are of equal length.
    /// Case Sensitive.
    /// </param>
    public WordSearch(string grid)
    {
        this.grid = grid.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
    }

    /// <summary>
    /// Return a dictionary of coordinates (if found) of a given 
    /// word in the word search. If not found the entry in the 
    /// dictionary should be null.
    /// </summary>
    /// <param name="wordsToSearchFor">Array of strings to search for (case-sensitive).</param>
    /// <returns>Dictionary where the key is the word searched for and the value
    /// is a null if not found otherwise a pair of coordinates for the start
    /// and end of the word in the grid expressed as a tuple of integer tuple pairs.</returns>
    public Dictionary<string, ((int, int), (int, int))?> Search(string[] wordsToSearchFor)
    {
        Dictionary<string, ((int, int), (int, int))?> ret = new();

        // Initialize the dictionary with nulls for each word.
        wordsToSearchFor.ForEach<string>(word => ret.Add(word, null));

        // We do the top-bottom with a transposed copy of the grid.
        string[]? transpose = (from int i in Enumerable.Range(0, grid[0].Length)
                               select new string((from row in grid
                                                  select row[i]).ToArray<char>())
            ).ToArray();

        // Get a list of possible diagonals so we can use index of instead of 
        // searching for letters individually 
        // (with the transposed grid it makes things more consistent too).
        List<List<Tuple<int, int>>> diagonals = Diagonals(grid.Length, grid[0].Length);

        foreach (string word in wordsToSearchFor)
        {
            string reversed = ReverseString(word);
            // left to right, or right to left each line.
            foreach ((int rowNum, string row) in grid.Index())
            {
                if (row.Length == 0)
                    continue;

                int index = row.IndexOf(word);
                if (index != -1) // -1 == Not found.
                {
                    ret[word] = ((index + 1, rowNum + 1), (index + word.Length, rowNum + 1));
                    break;
                }

                index = row.IndexOf(reversed);
                if (index != -1)
                {
                    ret[word] = ((index + reversed.Length, rowNum + 1), (index + 1, rowNum + 1));
                    break;
                }
            }
        }

        // Vertical
        if (transpose != null) 
        {
            foreach (string word in wordsToSearchFor)
            {
                // Skip if we already found one.
                if (ret[word] != null)
                    continue;

                string reversed = ReverseString(word);

                // top to bottom, or bottom to top each line.
                foreach ((int colNum, string col) in transpose.Index())
                {
                    if (col.Length == 0)
                        continue;
                    int index = col.IndexOf(word);
                    if (index != -1) // -1 == Not found.
                    {
                        ret[word] = ((colNum + 1, index + 1), (colNum + 1, index + word.Length));
                        break;
                    }

                    index = col.IndexOf(reversed);
                    if (index != -1) // -1 == Not found.
                    {
                        ret[word] = ((colNum + 1, index + reversed.Length), (colNum + 1, index + 1));
                        break;
                    }
                }
            }
        }
        // Diagonals
        foreach (string word in wordsToSearchFor)
        {
            // Skip if we already found one.
            if (ret[word] != null)
                continue;

            foreach (List<Tuple<int, int>> diagonal in diagonals)
            {
                // No use searching if the word won't fit.
                if (diagonal.Count < word.Length)
                    continue;
                    
                // Convert diagonal list of indexes into a string.
                string? row = new string((from pair in diagonal
                                          select grid[pair.Item2][pair.Item1]).ToArray<char>());
                string reversed = ReverseString(word);
                if (row != null)
                {
                    int index = row.IndexOf(word);
                    if (index != -1) // -1 == Not found.
                    {
                        ret[word] = ((diagonal[index].Item1 + 1, diagonal[index].Item2 + 1), (diagonal[index + word.Length - 1].Item1 + 1, diagonal[index + word.Length - 1].Item2 + 1));
                        break;
                    }

                    index = row.IndexOf(reversed);
                    if (index != -1)
                    {
                        ret[word] = ((diagonal[index + word.Length - 1].Item1 + 1, diagonal[index + word.Length - 1].Item2 + 1), (diagonal[index].Item1 + 1, diagonal[index].Item2 + 1));
                        break;
                    }
                }
            }
        }
        return ret;
    }

    /// <summary>
    /// Calculate the diagonals of a given grid size.
    /// </summary>
    /// <param name="rows">Rows in the grid (y)</param>
    /// <param name="cols">Columns in the grid (x)</param>
    /// <returns>A list each diagonal where each 
    /// diagonal is a list of (int x, int y) coordinates
    /// </returns>
    public List<List<Tuple<int, int>>> Diagonals(int rows, int cols)
    {
        List<List<Tuple<int, int>>> ret = new();
        // Diagonal left to right.
        for (int j = 0; j < rows; j++)
        {
            List<Tuple<int, int>> diagonal = new();
            int x = 0;
            int y = j;
            while (x < cols && y < rows)
            {
                diagonal.Add(new Tuple<int, int>(x, y));
                x++; y++;
            }
            ret.Add(diagonal);
        }
        for (int i = 1; i < cols; i++)
        {
            List<Tuple<int, int>> diagonal = new();
            int x = i;
            int y = 0;
            while (x < cols && y < rows)
            {
                diagonal.Add(new Tuple<int, int>(x, y));
                x++; y++;
            }
            ret.Add(diagonal);
        }

        // Diagonal Right to Left
        for (int j = 0; j < rows; j++)
        {
            List<Tuple<int, int>> diagonal = new();
            int x = 0;
            int y = j;
            while (x < cols && y >= 0)
            {
                diagonal.Add(new Tuple<int, int>(x, y));
                x++; y--;
            }
            ret.Add(diagonal);
        }
        for (int i = 1; i < cols; i++)
        {
            List<Tuple<int, int>> diagonal = new();
            int x = i;
            int y = rows - 1;
            while (x < cols && y >= 0)
            {
                diagonal.Add(new Tuple<int, int>(x, y));
                x++; y--;
            }
            ret.Add(diagonal);
        }
        return ret;
    }

    /// <summary>
    /// Return a reversed copy of a given string
    /// </summary>
    /// <param name="value">The string to return a reversed copy of.</param>
    /// <returns>The given value, but reversed</returns>
    string ReverseString(string value)
    {
        return new string(value.Reverse().ToArray());
    }
}

File: go/word-search/word_search.go
// Word Search Exercise
package wordsearch

import (
	"errors"
	"strings"
)

// Information about a diagonal on the puzzle
type DiagonalInfo struct {
	startX int
	startY int
	endX   int
	endY   int
	text   string
}

// Find the given words in the word search and return their locations.
// @param words: slice of string - the words to search for
// @param puzzle: slice of string - rows of the puzzle.
// @returns: map of word to x, y location of start and end of the word in the puzzle
func Solve(words []string, puzzle []string) (map[string][2][2]int, error) {
	result := make(map[string][2][2]int, 0)

	for _, word := range words {
		result[word] = [2][2]int{{-1, -1}, {-1, -1}} // default
	}

	for _, word := range words {
		wordReverse := ReverseString(word)
		found := false

		for rowIndex, row := range puzzle {
			// Look for word left to right or right to left.
			leftToRight := strings.Index(row, word)
			rightToLeft := strings.Index(row, wordReverse)
			if leftToRight >= 0 {
				result[word] = [2][2]int{{leftToRight, rowIndex}, {leftToRight + len(word) - 1, rowIndex}}
				found = true
				break
			} else if rightToLeft >= 0 {
				result[word] = [2][2]int{{rightToLeft + len(word) - 1, rowIndex}, {rightToLeft, rowIndex}}
				found = true
				break
			}
		}
		if !found && len(words) > 0 {
			// Now try columns
			for colIndex := 0; colIndex < len(puzzle[0]); colIndex++ {
				//Build the column
				column := ""
				for rowIndex := 0; rowIndex < len(puzzle); rowIndex++ {
					column += string(puzzle[rowIndex][colIndex])
				}

				// Look for word top to bottom or bottom to top.
				topToBottom := strings.Index(column, word)
				bottomToTop := strings.Index(column, wordReverse)
				if topToBottom >= 0 {
					result[word] = [2][2]int{{colIndex, topToBottom}, {colIndex, topToBottom + len(word) - 1}}
					found = true
					break
				} else if bottomToTop >= 0 {
					result[word] = [2][2]int{{colIndex, bottomToTop + len(word) - 1}, {colIndex, bottomToTop}}
					found = true
					break
				}
			}
		}

		if !found && len(words) > 0 {
			// Now try diagonals
			diagonals := Diagonals(puzzle)
			for _, diagonal := range diagonals {

				// Look for word top to bottom or bottom to top.
				forward := strings.Index(diagonal.text, word)
				reverse := strings.Index(diagonal.text, wordReverse)
				dx := 1
				if diagonal.startX > diagonal.endY {
					dx = -1
				}
				dy := 1
				if diagonal.startY > diagonal.endY {
					dy = -1
				}
				if forward >= 0 {
					start := diagonal.startX + (dx * forward)
					end := diagonal.startY + (dy * forward)
					result[word] = [2][2]int{{start, end}, {start + (dx * (len(word) - 1)), end + (dy * (len(word) - 1))}}
					found = true
					break
				} else if reverse >= 0 {
					start := diagonal.startX + (dx * (reverse + len(word) - 1))
					end := diagonal.startY + (dy * (reverse + len(word) - 1))
					result[word] = [2][2]int{{start, end}, {start - (dx * (len(word) - 1)), end - (dy * (len(word) - 1))}}
					found = true
					break
				}
			}
		}

		if !found {
			return result, errors.New("Word not found: " + word)
		}
	}
	// list 2,5-> 5,2 instead of 2,9 -> 5,12
	return result, nil
}

// Reverse a string, why is this not in the GO common libraries?
// @param line: the string to reverse
// @returns: reversed copy of the input string. May fail on multi-byte strings.
func ReverseString(line string) string {
	result := make([]rune, len(line))
	lastIndex := len(line) - 1
	for i, char := range line {
		result[lastIndex-i] = char
	}
	return string(result)
}

// Calculate the diagonals of a given grid with a given size.
// @param grid: The grid to calculate the diagonals from.
// @returns: A slice of DiagonalInfo one for each diagonal (has start x,y end x,y and the text.)
func Diagonals(grid []string) []DiagonalInfo {
	rows := len(grid)
	cols := 0
	if rows > 0 {
		cols = len(grid[0])
	}
	result := make([]DiagonalInfo, 0)

	// Diagonal left to right.
	for j := 0; j < rows; j++ {
		var diagonal string = ""
		x := 0
		y := j
		for x < cols && y < rows {
			diagonal += string(grid[y][x])
			x++
			y++
		}
		result = append(result, DiagonalInfo{
			startX: 0, startY: j, endX: x - 1, endY: y - 1, text: diagonal})
	}
	for i := 1; i < cols; i++ {
		var diagonal string = ""
		x := i
		y := 0
		for x < cols && y < rows {
			diagonal += string(grid[y][x])
			x++
			y++
		}
		result = append(result, DiagonalInfo{
			startX: i, startY: 0, endX: x - 1, endY: y - 1, text: diagonal})
	}

	// Diagonal Right to Left
	for j := 0; j < rows; j++ {
		var diagonal string = ""
		x := 0
		y := j
		for x < cols && y >= 0 {
			diagonal += string(grid[y][x])
			x++
			y--
		}
		result = append(result, DiagonalInfo{startX: 0, startY: j, endX: x - 1, endY: y + 1, text: diagonal})
	}

	for i := 1; i < cols; i++ {
		var diagonal string = ""
		x := i
		y := rows - 1
		for x < cols && y >= 0 {
			diagonal += string(grid[y][x])
			x++
			y--
		}
		result = append(result, DiagonalInfo{startX: i, startY: rows - 1, endX: x - 1, endY: y + 1, text: diagonal})
	}
	return result
}
